src/Microsoft.AspNetCore.Razor.Language/Legacy/TokenizerBackedParser.cs

```diff
@@ -570,108 +492,202 @@ protected SyntaxToken AcceptWhiteSpaceInLines()
         }
 
-        private void Configure(SpanKindInternal? kind, AcceptedCharactersInternal? accepts)
+        protected internal bool Balance(SyntaxListBuilder<RazorSyntaxNode> builder, BalancingModes mode, SyntaxKind left, SyntaxKind right, SourceLocation start)
         {
-            if (kind != null)
+            var startPosition = CurrentStart.AbsoluteIndex;
+            var nesting = 1;
+            if (!EndOfFile)
             {
-                Span.Kind = kind.Value;
+                var tokens = new List<SyntaxToken>();
+                do
+                {
+                    if (IsAtEmbeddedTransition(
+                        (mode & BalancingModes.AllowCommentsAndTemplates) == BalancingModes.AllowCommentsAndTemplates,
+                        (mode & BalancingModes.AllowEmbeddedTransitions) == BalancingModes.AllowEmbeddedTransitions))
+                    {
+                        Accept(tokens);
+                        tokens.Clear();
+                        ParseEmbeddedTransition(builder);
+
+                        // Reset backtracking since we've already outputted some spans.
+                        startPosition = CurrentStart.AbsoluteIndex;
+                    }
+                    if (At(left))
+                    {
+                        nesting++;
+                    }
+                    else if (At(right))
+                    {
+                        nesting--;
+                    }
+                    if (nesting > 0)
+                    {
+                        tokens.Add(CurrentToken);
+                    }
+                }
+                while (nesting > 0 && NextToken());
+
+                if (nesting > 0)
+                {
+                    if ((mode & BalancingModes.NoErrorOnFailure) != BalancingModes.NoErrorOnFailure)
+                    {
+                        Context.ErrorSink.OnError(
+                            RazorDiagnosticFactory.CreateParsing_ExpectedCloseBracketBeforeEOF(
+                                new SourceSpan(start, contentLength: 1 /* { OR } */),
+                                Language.GetSample(left),
+                                Language.GetSample(right)));
+                    }
+                    if ((mode & BalancingModes.BacktrackOnFailure) == BalancingModes.BacktrackOnFailure)
+                    {
+                        Context.Source.Position = startPosition;
+                        NextToken();
+                    }
+                    else
+                    {
+                        Accept(tokens);
+                    }
+                }
+                else
+                {
+                    // Accept all the tokens we saw
+                    Accept(tokens);
+                }
             }
-            if (accepts != null)
+            return nesting == 0;
+        }
+
+        protected virtual void ParseEmbeddedTransition(in SyntaxListBuilder<RazorSyntaxNode> builder)
```

Comment:
No need to be virtual, can inline the implementation into Balance when it's moved into CSharpCodeParser or, if the implementation is massive can just leave the method in the CSharpCodeParser

Reply:
Ditto with IsAtEmbeddedTransition